File: leetcode/13.Roman-To-Integer/RomanToInteger.cpp
#include <iostream>
#include <unordered_map>
#include <string>

using namespace std;

class RomanToInteger
{
public:
    int romanToInt(string s)
    {
        unordered_map<char, int> romanMap = {{'I', 1}, {'V', 5}, {'X', 10}, {'L', 50}, {'C', 100}, {'D', 500}, {'M', 1000}};
        int total = romanMap[s[s.size() - 1]];
        for (int i = s.size() - 2; i >= 0; i--)
        {
            int value = romanMap[s[i]];
            if (value < romanMap[s[i + 1]])
            {
                total -= value;
            }
            else
            {
                total += value;
            }
        }
        return total;
    }
};

File: leetcode/13.Roman-To-Integer/RomanToInteger.java
import java.util.HashMap;

public class RomanToInteger {

    public static int romanToInt(String s) {
        HashMap<Character, Integer> values = new HashMap<>();
        values.put('I', 1);
        values.put('V', 5);
        values.put('X', 10);
        values.put('L', 50);
        values.put('C', 100);
        values.put('D', 500);
        values.put('M', 1000);

        char[] letters = s.toCharArray();
        int i = 0;
        int res = 0;
        while (i < letters.length) {
            int c1 = values.get(letters[i]);
            if (i == letters.length - 1) {
                res += c1;
                break;
            }

            int c2 = values.get(letters[++i]);

            if (c1 >= c2) {
                res += c1;
            } else {
                res += c2 - c1;
                i++;
            }
        }

        return res;
    }
}